src/whitelist.hpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022, KNS Group LLC (YADRO).
+ */
+#pragma once
+
+#include <filesystem>
+#include <set>
+
+namespace fs = std::filesystem;
+
+namespace fssync
+{
+
+namespace details
+{
+struct Pred
+{
+    /**
+     * @brief Compare the begin of strings.
+     *
+     * For example:
+     * The comparison of the strings like `etc/systemd` and
+     * `etc/systemd/network` should return true.
+     */
+    bool operator()(const std::string& lhs, const std::string& rhs) const
+    {
+        const auto llen = lhs.length();
+        const auto rlen = rhs.length();
+        return lhs.compare(0, llen, rhs, 0, std::min(llen, rlen)) < 0;
```

Comment:
I guess with C++20 you could use just `rhs.starts_with(lhs)` ?

Reply:
Until we catch up the upstream in our internal repo we have to use C++17.  And this code is designed exactly for C++17.